source/opt/loop_descriptor.h

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_LOOP_DESCRIPTORS_H_
+#define LIBSPIRV_OPT_LOOP_DESCRIPTORS_H_
+
+#include <algorithm>
+#include <cstdint>
+#include <map>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "opt/module.h"
+#include "opt/pass.h"
+#include "opt/tree_iterator.h"
+
+namespace spvtools {
+namespace ir {
+class CFG;
+class LoopDescriptor;
+
+// A class to represent and manipulate a loop.
+class Loop {
+  // The type used to represent nested child loops.
+  using ChildrenList = std::vector<Loop*>;
+
+ public:
+  using iterator = ChildrenList::iterator;
+  using const_iterator = ChildrenList::const_iterator;
+  using BasicBlockListTy = std::unordered_set<uint32_t>;
+
+  Loop()
+      : loop_header_(nullptr),
+        loop_continue_(nullptr),
+        loop_merge_(nullptr),
+        loop_preheader_(nullptr),
+        parent_(nullptr) {}
+
+  Loop(IRContext* context, opt::DominatorAnalysis* analysis, BasicBlock* header,
+       BasicBlock* continue_target, BasicBlock* merge_target);
+
+  // Iterators which allows access to the nested loops.
+  inline iterator begin() { return nested_loops_.begin(); }
+  inline iterator end() { return nested_loops_.end(); }
+  inline const_iterator begin() const { return cbegin(); }
+  inline const_iterator end() const { return cend(); }
+  inline const_iterator cbegin() const { return nested_loops_.begin(); }
+  inline const_iterator cend() const { return nested_loops_.end(); }
+
+  // Returns the header (first basic block of the loop). This block contains the
+  // OpLoopMerge instruction.
+  inline BasicBlock* GetHeaderBlock() { return loop_header_; }
+  inline const BasicBlock* GetHeaderBlock() const { return loop_header_; }
+
+  // Returns the latch basic block (basic block that holds the back-edge).
+  inline BasicBlock* GetLatchBlock() { return loop_continue_; }
+  inline const BasicBlock* GetLatchBlock() const { return loop_continue_; }
+
+  // Returns the BasicBlock which marks the end of the loop.
+  inline BasicBlock* GetMergeBlock() { return loop_merge_; }
+  inline const BasicBlock* GetMergeBlock() const { return loop_merge_; }
```

Comment:
So, this will not work with compute kernels?  If it does, please document what happens when this function is used in unstructured IR.

Reply:
This did not seem to get addressed.